server/coordinator/procedure/manager_impl.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
+
+package procedure
+
+import (
+	"context"
+	"sync"
+
+	"github.com/CeresDB/ceresmeta/pkg/log"
+	"github.com/CeresDB/ceresmeta/server/cluster"
+	"github.com/CeresDB/ceresmeta/server/coordinator/eventdispatch"
+	"github.com/pkg/errors"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.uber.org/zap"
+)
+
+const (
+	queueSize         = 10
+	metaListBatchSize = 100
+)
+
+type ManagerImpl struct {
+	// This lock is used to protect the field `procedures`.
+	lock       sync.RWMutex
+	procedures []Procedure
+
+	storage  Storage
+	dispatch eventdispatch.Dispatch
+
+	procedureQueue   chan Procedure
+	resultChannelMap map[uint64]chan error
+}
+
+func (m *ManagerImpl) Start(ctx context.Context) error {
+	m.procedureQueue = make(chan Procedure, queueSize)
+	m.resultChannelMap = make(map[uint64]chan error, 0)
+	go m.startProcedureWorker(ctx, m.procedureQueue)
+	err := m.retryAll(ctx)
+	if err != nil {
+		return errors.WithMessage(err, "retry all running procedure failed")
+	}
+	return nil
+}
+
+func (m *ManagerImpl) Stop(ctx context.Context) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	close(m.procedureQueue)
+	for _, procedure := range m.procedures {
+		if procedure.State() == StateRunning {
+			err := procedure.Cancel(ctx)
+			log.Error("cancel procedure failed", zap.Error(err), zap.Uint64("procedureID", procedure.ID()))
+			// TODO: consider whether a single procedure cancel failed should return directly.
+			return err
+		}
+	}
+	return nil
+}
+
+func (m *ManagerImpl) Submit(_ context.Context, procedure Procedure) (<-chan error, error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	resultChannel := make(chan error, 1)
+	m.resultChannelMap[procedure.ID()] = resultChannel
```

Comment:
No need to suffix the type:
```suggestion
	m.resultChannels[procedure.ID()] = resultChannel
```

Btw, when to clear the unused result channel?